### PR TITLE
fix: Add check for shorebird yaml file in run command

### PR DESCRIPTION
**Pull request by rkishan516**
## Description

I have added shorebird missing file check for run and build command as asked by @eseidel   at issue #414.
After this we will check for shorebird file just after authentication check in run and builds command's

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/shorebird_cli/lib/src/commands/build/build_apk_command.dart [hunk 1]
```diff
@@ -49,6 +49,13 @@ class BuildApkCommand extends ShorebirdCommand
       return ExitCode.noUser.code;
     }
 
+    if (!hasShorebirdYaml) {
+      logger.err(
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/build/build_apk_command.dart]: I wonder whether this should be another validation check? cc @bryanoltman

Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/commands/build/build_apk_command.dart]: I think that's a great idea

Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/build/build_apk_command.dart]: Sure, Will update check with validation.

**Comment by rkishan516** [comment]
@bryanoltman @felangel I have added validator for checking shorebird yaml.

**Comment by rkishan516** [comment]
@bryanoltman @felangel any update, Should I added tests here ?

**Comment by bryanoltman** [comment]
We'll definitely need tests. It also looks like a few of the existing tests are broken (see https://github.com/shorebirdtech/shorebird/actions/runs/4966615389/jobs/8888149349?pr=451)

**Comment by rkishan516** [comment]
> We'll definitely need tests. It also looks like a few of the existing tests are broken (see https://github.com/shorebirdtech/shorebird/actions/runs/4966615389/jobs/8888149349?pr=451)

Yaa, I just wanted this to get reviewed.

**Comment by rkishan516** [comment]
@bryanoltman I have added all required test + fixed failing tests.
Sorry for the delay.

**Comment by rkishan516** [comment]
@bryanoltman @felangel can we take any action on this MR ?

File: packages/shorebird_cli/lib/src/validators/shorebird_yaml_validator.dart [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import 'package:shorebird_cli/src/shorebird_process.dart';
+import 'package:shorebird_cli/src/validators/validators.dart';
+
+class ShorebirdYamlValidator extends Validator {
+  ShorebirdYamlValidator({required this.hasShorebirdYaml});
+  // Adding as Function
+  /// Returns true if the project has a shorebird.yaml file.
+  final bool Function() hasShorebirdYaml;
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/validators/shorebird_yaml_validator.dart]: Why are we injecting a function to do the validation? Thoughts on performing the validation directly in the validator? We can still unit test it and it removes one level of indirection imo. Let me know what you think!

File: packages/shorebird_cli/lib/src/commands/run_command.dart [hunk 3]
```diff
@@ -1,10 +1,13 @@
 import 'dart:convert';
 
+import 'package:collection/collection.dart';
 import 'package:mason_logger/mason_logger.dart';
 import 'package:shorebird_cli/src/auth_logger_mixin.dart';
 import 'package:shorebird_cli/src/command.dart';
 import 'package:shorebird_cli/src/shorebird_config_mixin.dart';
 import 'package:shorebird_cli/src/shorebird_validation_mixin.dart';
+import 'package:shorebird_cli/src/validators/shorebird_yaml_validator.dart';
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/run_command.dart]:
```suggestion
```
Minor but we should update the barrel file to include the `shorebird_yaml_validator.dart` and this additional import won't be necessary

File: packages/shorebird_cli/lib/src/commands/build/build_command.dart [hunk 4]
```diff
@@ -1,21 +1,47 @@
+import 'package:collection/collection.dart';
 import 'package:shorebird_cli/src/command.dart';
 import 'package:shorebird_cli/src/commands/build/build.dart';
+import 'package:shorebird_cli/src/shorebird_config_mixin.dart';
+import 'package:shorebird_cli/src/validators/shorebird_yaml_validator.dart';
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/build/build_command.dart]:
```suggestion
```

Same here

File: packages/shorebird_cli/lib/src/commands/build/build_command.dart [hunk 5]
```diff
@@ -1,21 +1,47 @@
...
+class BuildCommand extends ShorebirdCommand with ShorebirdConfigMixin {
   /// {@macro build_command}
   BuildCommand({required super.logger}) {
+    validators = _allValidators(baseValidators: validators);
     addSubcommand(BuildApkCommand(logger: logger));
     addSubcommand(BuildAppBundleCommand(logger: logger));
     addSubcommand(BuildIpaCommand(logger: logger));
   }
 
+  late final List<Validator> _buildValidators = [
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/build/build_command.dart]: Why is this `late`?

File: packages/shorebird_cli/lib/src/commands/build/build_command.dart [hunk 6]
```diff
@@ -1,21 +1,47 @@
...
+    ShorebirdYamlValidator(hasShorebirdYaml: () => hasShorebirdYaml),
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/build/build_command.dart]:
```suggestion
    ShorebirdYamlValidator(),
```

File: packages/shorebird_cli/lib/src/commands/run_command.dart [hunk 7]
```diff
@@ -83,4 +93,20 @@ class RunCommand extends ShorebirdCommand
 
     return flutter.exitCode;
   }
+
+  /// Creates a list that is the union of [baseValidators] and
+  /// [_runValidators].
+  List<Validator> _allValidators({
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/run_command.dart]: Can we avoid duplicating this code? I think it'd be nice to refactor `validators` to be a `Set` and each command can add it's own custom validators to the default set without having to compute the union like this. Thoughts?

File: packages/shorebird_cli/lib/src/validators/shorebird_yaml_validator.dart (hunk 2 above)
Comment (rkishan516) [reply packages/shorebird_cli/lib/src/validators/shorebird_yaml_validator.dart]: Yup, make sense. I think we should do it directly inside validator. I'll update

File: packages/shorebird_cli/lib/src/commands/run_command.dart (hunk 7 above)
Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/run_command.dart]: Yaa, making it a `Set` will solve this. Also we have validators on command level only. So, I think we should do this change.

Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/run_command.dart]: I'll update it accordingly

File: packages/shorebird_cli/lib/src/commands/build/build_command.dart (hunk 5 above)
Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/build/build_command.dart]: My bad, should not be late.

File: packages/shorebird_cli/lib/src/commands/build/build_command.dart (hunk 4 above)
Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/build/build_command.dart]: Will update it

File: packages/shorebird_cli/lib/src/commands/run_command.dart (hunk 3 above)
Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/run_command.dart]: Will update it

File: packages/shorebird_cli/lib/src/commands/run_command.dart (hunk 7 above)
Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/run_command.dart]: @felangel Have updated to Set. but `ensure_build` test is failing. Can you check that ?

Comment (felangel) [reply packages/shorebird_cli/lib/src/commands/run_command.dart]: Did you re-run `build_runner`?

Comment (rkishan516) [reply packages/shorebird_cli/lib/src/commands/run_command.dart]: Yes, I did.

**Comment by felangel** [comment]
Closing this since run is deprecated, thanks!

**Pull request closed without merging**